src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs

```diff
@@ -0,0 +1,177 @@
+﻿namespace ServiceControl.Transports.RabbitMQ
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using global::RabbitMQ.Client;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.Messaging;
+    using Monitoring.QueueLength;
+    using NServiceBus.Logging;
+    using NServiceBus.Metrics;
+
+    public class QueueLengthProvider : IProvideQueueLength
+    {
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            queryExecutor = new QueryExecutor(connectionString);
+
+            this.store = store;
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            var endpointInstanceQueue = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+            var queueName = metadataReport.LocalAddress;
+
+            endpointQueues.AddOrUpdate(endpointInstanceQueue, _ => queueName, (_, currentValue) =>
+            {
+                if (currentValue != queueName) sizes.TryRemove(currentValue, out var _);
+
+                return queueName;
+            });
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+            //RabbitMQ does not support endpoint level queue length reports
+        }
+
+        public Task Start()
+        {
+            queryExecutor.Initialize();
+
+            queryLoop = Task.Run(async () =>
+            {
+                while (!stoppedTokenSource.Token.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await FetchQueueLengths();
+
+                        UpdateQueueLengths();
+
+                        await Task.Delay(QueryDelayInterval).ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Error("Queue length query loop failure.", e);
+                    }
+                }
+            });
+
+            return TaskEx.Completed;
+        }
+
+        public Task Stop()
+        {
+            stoppedTokenSource.Cancel();
+
+            return queryLoop;
+        }
+
+        void UpdateQueueLengths()
+        {
+            var nowTicks = DateTime.UtcNow.Ticks;
+
+            foreach (var endpointQueuePair in endpointQueues)
+                store.Store(
+                    new[]
+                    {
+                        new RawMessage.Entry
+                        {
+                            DateTicks = nowTicks,
+                            Value = sizes.TryGetValue(endpointQueuePair.Value, out var size) ? size : 0
```

Comment:
Is it correct to report 0 where we have no measurement?

Reply:
This is a good point.